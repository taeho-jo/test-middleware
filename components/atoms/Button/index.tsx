import React, { ButtonHTMLAttributes } from 'react';
// Components
import ClipLoader from 'react-spinners/ClipLoader';
// Styles
import { css } from '@emotion/react';
import { commonStyles } from '../../../styles/Common.styles';
const { colors } = commonStyles;
// Types
interface PropsType extends ButtonHTMLAttributes<HTMLButtonElement> {
  children?: string | React.ReactNode;
  size?: 'sm' | 'md' | 'lg';
  bgColor?: 'primary' | 'secondary' | 'gray' | 'black' | 'danger';
  loading?: boolean;
  active?: boolean;
  type?: 'button' | 'submit' | 'reset';
  border?: 'none' | 'outline' | 'text';
}

const ButtonSizeStyle = {
  sm: { width: '60px', padding: '8px 0' },
  md: { width: '80px', padding: '8px 0' },
  lg: { width: '100px', padding: '8px 0' },
};

const Button = ({
  children = 'Button',
  size = 'sm',
  bgColor = 'primary',
  loading = false,
  active = true,
  type = 'button',
  border = 'none',
}: PropsType) => {
  return loading ? (
    <button type={type} css={test(size, bgColor, active, border)}>
      <ClipLoader
        color={border === 'none' ? colors.wh_1 : colors[bgColor]}
        loading={loading}
        size={16}
      />
    </button>
  ) : (
    <button css={test(size, bgColor, active, border)}>{children}</button>
  );
};

export default Button;

const test = (size, bgColor, active, border) => css`
  width: ${ButtonSizeStyle[size].width};
  padding: ${ButtonSizeStyle[size].padding};
  background: ${active && border === 'none'
    ? colors[bgColor]
    : active && (border === 'outline' || border === 'text')
    ? '#fff'
    : colors.gray};
  border: ${border === 'none' || border === 'text'
    ? 'none'
    : `2px solid ${colors[bgColor]}`};
  border-radius: 8px;
  color: ${border === 'none' ? colors.wh_1 : colors[bgColor]};
  cursor: ${active ? 'pointer' : 'not-allowed'};
  font-weight: bold;
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 4px;
  ${active ? `&:hover { background: ${colors[`hover_${bgColor}`]};}` : ''}
`;
