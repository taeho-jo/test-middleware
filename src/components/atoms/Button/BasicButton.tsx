import React, { ButtonHTMLAttributes } from 'react';
import { css } from '@emotion/react';
import { colors } from '../../../styles/Common.styles';
import { fontStyle } from '@mui/system';
import { heading4_bold } from '../../../styles/FontStyles';
import ClipLoader from 'react-spinners/ClipLoader';

interface PropsType extends ButtonHTMLAttributes<HTMLButtonElement> {
  isLoading?: boolean;
  designBgColor?: string;
  text?: string;
  type?: 'submit' | 'button' | 'reset';
  status?: 'normal' | 'disabled';
  theme?: 'light' | 'dark';
  style?: any;
  onClick?: () => void;
  [key: string]: any;
}

const BasicButton = ({
  isLoading = false,
  style,
  designBgColor,
  text,
  type = 'button',
  status = 'normal',
  theme = 'light',
  onClick,
  ...props
}: PropsType) => {
  const buttonTextStyle = heading4_bold;
  return (
    <button
      {...props}
      onClick={onClick}
      type={type}
      disabled={isLoading}
      css={[buttonStyle(status, theme, isLoading), { ...style }, { background: designBgColor ? designBgColor : '' }]}
    >
      {isLoading ? (
        <span css={[buttonTextStyle, textStyle]}>
          <ClipLoader size={10} color={'white'} />
        </span>
      ) : (
        <span css={[buttonTextStyle, textStyle]}>{text ? text : 'button'}</span>
      )}
    </button>
  );
};

export default BasicButton;

const buttonStyle = (status, theme, isLoading) => css`
  width: 100%;
  padding: 16px 56px;
  border-radius: 8px;
  border: none;
  box-sizing: border-box;
  cursor: ${status === 'disabled' || isLoading ? 'not-allowed' : 'pointer'};
  ${theme === 'light'
    ? `
    ${
      status === 'normal'
        ? `
      background: ${colors.cyan._500};
      &:hover {
        background: ${colors.cyan._700};
      }
    `
        : `
          background: ${colors.cyan._300};
        `
    };
  `
    : `
    ${
      status === 'normal'
        ? `
      background: ${colors.grey._3c};
      &:hover {
        background: ${colors.grey._2c};
      }
    `
        : `
        background: ${colors.grey._cc};
        `
    }
  `}
`;
const textStyle = css`
  color: white;
`;
