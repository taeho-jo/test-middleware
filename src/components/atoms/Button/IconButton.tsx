import React from 'react';
import Icon from '../Icon';
import { IconType } from '../../../common/types/commonTypes';
import { css } from '@emotion/react';

interface PropsType {
  type?: 'submit' | 'button' | 'reset';
  name: IconType;
  size?: number;
  style?: any;
  onClick?: () => void;
  [key: string]: any;
}

const IconButton = ({ type = 'button', name, size = 24, onClick, style, ...props }: PropsType) => {
  return (
    <button type={type} css={iconButtonStyle} {...props}>
      <Icon onClick={onClick} name={name} size={size} style={style} />
    </button>
  );
};

export default IconButton;

const iconButtonStyle = css`
  background: none;
  border: none;
  cursor: pointer;
  margin: 0;
  padding: 0;
`;
