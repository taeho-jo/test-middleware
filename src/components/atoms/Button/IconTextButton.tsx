import React from 'react';
import Icon from '../Icon';
import { heading5_regular } from '../../../styles/FontStyles';
import { IconType } from '../../../common/types/commonTypes';
import { css } from '@emotion/react';
import { colors } from '../../../styles/Common.styles';

interface PropsType {
  // status?: 'normal' | 'disabled';
  type?: 'submit' | 'button' | 'reset';
  text?: string;
  textStyle?: any;
  name: IconType;
  iconPosition?: 'left' | 'right';
  roundBorder?: boolean;
  onClick?: () => void;
  disabled?: boolean;
  [key: string]: any;
}

const IconTextButton = ({
  // status = 'disabled',
  textStyle = heading5_regular,
  text = 'Button',
  type = 'button',
  name,
  iconPosition = 'right',
  roundBorder = true,
  onClick,
  disabled = false,
  ...props
}: PropsType) => {
  return (
    <button disabled={disabled} type={type} onClick={onClick} css={iconTextButtonStyle(roundBorder, iconPosition, status)} {...props}>
      {iconPosition === 'left' ? <Icon name={disabled ? `${name}_DISABLED` : name} size={24} /> : null}
      <span
        css={[
          textStyle === 'custom' ? { fontSize: '14px', fontWeight: '400' } : textStyle,
          noBorderTextStyle(roundBorder, disabled),
          { margin: iconPosition === 'right' ? '0 8px 0 0' : '0 0 0 8px' },
        ]}
      >
        {text}
      </span>
      {iconPosition === 'right' ? <Icon name={disabled ? `${name}_DISABLED` : name} size={24} /> : null}
    </button>
  );
};

export default IconTextButton;

const iconTextButtonStyle = (roundBorder, iconPosition, status) => css`
  display: flex;
  align-items: center;
  box-sizing: border-box;
  justify-content: space-between;
  cursor: pointer;
  padding: ${iconPosition === 'right' ? '0 8px 0 16px' : '0 16px 0 8px'};
  ${roundBorder
    ? `
    border: 1px solid ${colors.grey._3c};
    border-radius: 36px;
    background: ${colors.white};
  `
    : `
    border: none;
    background: none;
  `}
  &:hover {
    ${roundBorder
      ? `
      background: ${colors.grey._ec};
    `
      : ``}
  }

  &:disabled {
    cursor: not-allowed;
    ${roundBorder
      ? `
      background: ${colors.grey._f7};
      border: 1px solid ${colors.grey._cc};
    `
      : ``}
  }
`;

const noBorderTextStyle = (roundBorder, disabled) => css`
  ${!roundBorder
    ? `
    &:hover {
      font-weight: 700;
    }
  `
    : ``}
  ${disabled
    ? `
    color: ${colors.grey._cc};
    
    &:hover {
      font-weight: 500;
    }
    
  `
    : ``}
`;
