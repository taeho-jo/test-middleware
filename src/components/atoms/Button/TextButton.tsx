import React from 'react';
import { css } from '@emotion/react';
import { caption1_regular, heading5_medium } from '../../../styles/FontStyles';

interface PropsType {
  type?: 'submit' | 'button' | 'reset';
  text?: string;
  textStyle: any;
  onClick?: () => void;
}

const TextButton = ({ type = 'button', text = 'button', textStyle = heading5_medium, onClick }: PropsType) => {
  return (
    <button type={type} onClick={onClick} css={textButtonStyle}>
      <span css={[textStyle, { textDecoration: 'underline' }]}>{text}</span>
    </button>
  );
};

export default TextButton;

const textButtonStyle = css`
  border: none;
  background: none;
  cursor: pointer;
`;
