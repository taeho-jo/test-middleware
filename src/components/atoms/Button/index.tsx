import React from 'react';
// Components
import Icon from '../Icon';
import ClipLoader from 'react-spinners/ClipLoader';
// Styles
import { css } from '@emotion/react';
import { colors } from '../../../styles/Common.styles';
import { body2_bold, caption1_regular } from '../../../styles/FontStyles';
// Types
import { ColorsType, IconType } from '../../../common/types/commonTypes';

interface PropsType {
  buttonType?: 'basic' | 'action' | 'icon';
  backgroundColor?: ColorsType;
  full?: boolean;
  type?: 'button' | 'submit' | 'reset';
  onClick?: () => void;
  btnText?: string;
  isLoading?: boolean;
  icon?: IconType;
  padding?: string;
  start?: string;
  size?: number;
  btnTextColor?: string;
  [key: string]: any;
}

const Button = ({
  buttonType = 'basic',
  type = 'button',
  backgroundColor = '#24e1d5',
  full = false,
  onClick,
  btnText = 'button',
  isLoading = false,
  icon,
  padding,
  start = '',
  size,
  btnTextColor,
  ...props
}: PropsType) => {
  return (
    <>
      {buttonType === 'basic' ? (
        <button type={type} onClick={onClick} css={buttonStyle(backgroundColor, full, padding)} {...props}>
          <div css={isLoading ? loadingStyle : ''}>
            <ClipLoader color={'white'} loading={isLoading} size={16} />
          </div>
          <span css={isLoading ? textStyle : [body2_bold, { color: btnTextColor }]}>{btnText}</span>
        </button>
      ) : buttonType === 'action' ? (
        <button type={type} onClick={onClick} css={actionButtonStyle(full, padding)}>
          {start === 'icon' ? (
            <>
              <Icon name={icon ? icon : 'NAVIGATION_CHEVRON_RIGHT'} size={size ? size : 24} />
              <span css={[caption1_regular, { marginLeft: '12px' }]}>{btnText}</span>
            </>
          ) : (
            <>
              <span css={caption1_regular}>{btnText}</span>
              <Icon name={icon ? icon : 'NAVIGATION_CHEVRON_RIGHT'} size={size ? size : 24} />
            </>
          )}
        </button>
      ) : (
        <button type={type} onClick={onClick} css={iconButtonStyle}>
          <Icon name={icon ? icon : 'NAVIGATION_CHEVRON_RIGHT'} size={size ? size : 24} />
          {/*<Icon name={'ACTION_CREATE'} size={20} />*/}
        </button>
      )}
    </>
  );
};

export default Button;

// basic button
const buttonStyle = (backgroundColor, full, padding) => css`
  padding: ${padding ? padding : '14px 53.5px'};
  background: ${backgroundColor ? backgroundColor : colors.cyan._500};
  border-radius: 12px;
  border: none;
  color: ${colors.white};
  width: ${full ? '100%' : 'auto'};
  cursor: pointer;
  position: relative;
  box-sizing: border-box;
`;

// action button
const actionButtonStyle = (full, padding) => css`
  padding: ${padding ? padding : '5px 7px 5px 16px'};
  background: ${colors.white};
  border: 1px solid ${colors.grey._3c};
  border-radius: 58px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  cursor: pointer;
  width: ${full ? '100%' : 'auto'};
  box-sizing: border-box;
}
`;

// icon button
const iconButtonStyle = css`
  cursor: pointer;
  padding: 0;
  margin: 0;
  border-radius: 50%;
  border: 1px solid ${colors.grey._3c};
  background: ${colors.white};
  width: 20px;
  height: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
`;

// loading style
const loadingStyle = css`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  //z-index: 10;
  padding: 14px 53.5px;
  box-sizing: border-box;
`;

const textStyle = [
  body2_bold,
  css`
    visibility: hidden;
  `,
];
