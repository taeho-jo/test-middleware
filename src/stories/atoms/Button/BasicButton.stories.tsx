import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
// Components
import BasicButton from '../../../components/atoms/Button/BasicButton';
// Styles

export default {
  title: 'ATOMS/Buttons',
  component: BasicButton,
} as ComponentMeta<typeof BasicButton>;

const Template: ComponentStory<typeof BasicButton> = args => {
  return (
    <div style={{ padding: '10px' }}>
      <BasicButton {...args} />
    </div>
  );
};

export const BasicButtonComponent = Template.bind({});
BasicButtonComponent.args = {
  text: '간편하게 시작하기',
  onClick: () => {
    return;
  },
};
