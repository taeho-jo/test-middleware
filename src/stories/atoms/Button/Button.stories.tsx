import React from 'react';
import { ComponentMeta } from '@storybook/react';
// Components
import FlexBox from '../../../components/atoms/FlexBox';
import { body2_regular, heading1_bold, heading5_medium } from '../../../styles/FontStyles';
import { colors } from '../../../styles/Common.styles';
import { css } from '@emotion/react';
import BasicButton from '../../../components/atoms/Button/BasicButton';
import IconButton from '../../../components/atoms/Button/IconButton';
import TextButton from '../../../components/atoms/Button/TextButton';
import IconTextButton from '../../../components/atoms/Button/IconTextButton';
// Styles

export default {
  title: 'ATOMS/Buttons',
  component: BasicButton,
} as ComponentMeta<typeof BasicButton>;

export const AllButton = () => {
  const titleStyle = [
    heading1_bold,
    css`
      margin-bottom: 20px;
    `,
  ];

  const subTitleStyle = [
    body2_regular,
    css`
      display: block;
      margin-bottom: 10px;
    `,
  ];

  return (
    <>
      <FlexBox justify={'flex-start'} align={'flex-start'} direction={'column'}>
        <h2 css={titleStyle}>Basic Button</h2>
        <FlexBox justify={'flex-start'} align={'flex-start'} width={'100%'} direction={'column'}>
          <FlexBox justify={'flex-start'} align={'flex-start'}>
            <div css={{ flex: 1, padding: '10px' }}>
              <h3 css={subTitleStyle}>üìçlight Button</h3>
              <BasicButton theme={'light'} status={'normal'} />
            </div>

            <div css={{ flex: 1, padding: '10px' }}>
              <h3 css={subTitleStyle}>üìçlight Hover Button</h3>
              <BasicButton designBgColor={colors.cyan._700} theme={'light'} status={'normal'} />
            </div>

            <div css={{ flex: 1, padding: '10px' }}>
              <h3 css={subTitleStyle}>üìçlight Disable Button</h3>
              <BasicButton theme={'light'} status={'disabled'} />
            </div>
          </FlexBox>

          <FlexBox justify={'flex-start'} align={'flex-start'} style={{ marginTop: '20px' }}>
            <div css={{ flex: 1, padding: '10px' }}>
              <h3 css={subTitleStyle}>üìçDark Button</h3>
              <BasicButton theme={'dark'} status={'normal'} />
            </div>

            <div css={{ flex: 1, padding: '10px' }}>
              <h3 css={subTitleStyle}>üìçDark Hover Button</h3>
              <BasicButton designBgColor={colors.grey._2c} theme={'dark'} status={'normal'} />
            </div>

            <div css={{ flex: 1, padding: '10px' }}>
              <h3 css={subTitleStyle}>üìçDark Disable Button</h3>
              <BasicButton theme={'dark'} status={'disabled'} />
            </div>
          </FlexBox>
        </FlexBox>
      </FlexBox>

      <FlexBox justify={'flex-start'} align={'flex-start'} direction={'column'} style={{ marginTop: '20px' }}>
        <h2 css={titleStyle}>Icon Button</h2>
        <FlexBox justify={'flex-start'} align={'flex-start'} wrap={'wrap'} width={'100%'}>
          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'GOOGLE'} />
            <p>GOOGLE</p>
          </FlexBox>
          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'ACTION_CREATE'} />
            <p>ACTION_CREATE</p>
          </FlexBox>

          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'ACTION_SETTING'} />
            <p>ACTION_SETTING</p>
          </FlexBox>

          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'ACTION_MORE'} />
            <p>ACTION_MORE</p>
          </FlexBox>

          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'NAVIGATION_ARROW_LEFT'} />
            <p>NAVIGATION_ARROW_LEFT</p>
          </FlexBox>

          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'NAVIGATION_ARROW_RIGHT'} />
            <p>NAVIGATION_ARROW_RIGHT</p>
          </FlexBox>

          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'NAVIGATION_CHEVRON_DOWN'} />
            <p>NAVIGATION_CHEVRON_DOWN</p>
          </FlexBox>

          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'NAVIGATION_CHEVRON_RIGHT'} />
            <p>NAVIGATION_CHEVRON_RIGHT</p>
          </FlexBox>

          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'NAVIGATION_CLOSE_LG'} />
            <p>NAVIGATION_CLOSE_LG</p>
          </FlexBox>

          <FlexBox justify={'center'} align={'center'} direction={'column'} style={{ padding: '10px', width: '20%' }}>
            <IconButton name={'NAVIGATION_CLOSE_SM'} />
            <p>NAVIGATION_CLOSE_SM</p>
          </FlexBox>
        </FlexBox>
      </FlexBox>

      <FlexBox justify={'flex-start'} align={'flex-start'} direction={'column'} style={{ marginTop: '20px' }}>
        <h2 css={titleStyle}>Text Button</h2>
        <TextButton textStyle={heading5_medium} text={'Îã§ÏùåÏóê Ìï†Í≤åÏöî'} />
      </FlexBox>

      <FlexBox justify={'flex-start'} align={'flex-start'} direction={'column'} style={{ marginTop: '20px' }}>
        <h2 css={titleStyle}>IconText Button</h2>

        <FlexBox justify={'flex-start'} align={'flex-start'}>
          <div css={{ flex: 1, padding: '10px' }}>
            <h3 css={subTitleStyle}>üìçBorder Right Button</h3>
            <IconTextButton disabled={false} name={'NAVIGATION_CHEVRON_RIGHT'} textStyle={heading5_medium} />
          </div>

          <div css={{ flex: 1, padding: '10px' }}>
            <h3 css={subTitleStyle}>üìçNo Border Right Button</h3>
            <IconTextButton disabled={false} name={'NAVIGATION_CHEVRON_RIGHT'} textStyle={heading5_medium} roundBorder={false} />
          </div>
        </FlexBox>
      </FlexBox>
    </>
  );
};
