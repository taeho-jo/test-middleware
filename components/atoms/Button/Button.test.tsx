import Button from './index';
import { render } from '@testing-library/react';
import React from 'react';

describe('atoms/Button', () => {
  it('matches snapshot', () => {
    const utils = render(
      <Button size={'sm'} bgColor={'primary'} loading={true} active={false} />,
    );
    expect(utils.container).toMatchSnapshot();
  });
});
