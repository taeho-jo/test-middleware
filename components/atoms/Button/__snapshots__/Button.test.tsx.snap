// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BUTTON!!!!!!!!!!!!!!!!! matches snapshot 1`] = `
<div>
  <button
    class="css-ljgqd9-test"
    type="button"
  >
    <span
      class="css-9q8ohv"
    />
  </button>
</div>
`;
